<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>LACHESIS by shendurelab</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1>LACHESIS</h1>
        <p>The LACHESIS software package, ultra-long-range scaffolding of de novo genome assemblies</p>

        <p class="view"><a href="https://github.com/shendurelab/LACHESIS">View the Project on GitHub <small>shendurelab/LACHESIS</small></a></p>


        <ul>
          <li><a href="https://github.com/shendurelab/LACHESIS/zipball/master">Download <strong>ZIP File</strong></a></li>
          <li><a href="https://github.com/shendurelab/LACHESIS/tarball/master">Download <strong>TAR Ball</strong></a></li>
          <li><a href="https://github.com/shendurelab/LACHESIS">View On <strong>GitHub</strong></a></li>
        </ul>
      </header>
      <section>
        <h3>
<a name="lachesis" class="anchor" href="#lachesis"><span class="octicon octicon-link"></span></a>LACHESIS.</h3>

<p>ABSTRACT:</p>

<p>Genomes assembled de novo from short reads are highly fragmented relative to the finished chromosomes of Homo sapiens and key model organisms generated by the Human Genome Project. To address this problem, we need scalable, cost-effective methods to obtain assemblies with chromosome-scale contiguity. Here we show that genome-wide chromatin interaction data sets, such as those generated by Hi-C, are a rich source of long-range information for assigning, ordering and orienting genomic sequences to chromosomes, including across centromeres. To exploit this finding, we developed an algorithm that uses Hi-C data for ultra-long-range scaffolding of de novo genome assemblies. We demonstrate the approach by combining shotgun fragment and short jump mate-pair sequences with Hi-C data to generate chromosome-scale de novo assemblies of the human, mouse and Drosophila genomes, achieving—for the human genome—98% accuracy in assigning scaffolds to chromosome groups and 99% accuracy in ordering and orienting scaffolds within chromosome groups. Hi-C data can also be used to validate chromosomal translocations in cancer genomes.</p>

<p>The full paper has been published in Nature Biotechnology, and is available through <a href="http://www.ncbi.nlm.nih.gov/pubmed/24185095">PubMed</a></p>
      </section>
      <footer>
        <p>This project is maintained by <a href="https://github.com/shendurelab">shendurelab</a></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></small></p>
      </footer>
    </div>
    <script src="javascripts/scale.fix.js"></script>
    
  </body>
</html>